src/Stratis.Bitcoin.Features.Miner/PosMinting.cs

```diff
@@ -446,17 +446,10 @@ public void StopStake()
                 return;
             }
 
-            this.stakeCancellationTokenSource.Cancel();
-            this.logger.LogTrace("Waiting for staking loop to cancel...");
-            this.stakingLoop?.RunningTask?.GetAwaiter().OnCompleted(() =>
-            {
-                this.logger.LogTrace("Staking loop cancelled.");                              
-            });
+            this.stakeCancellationTokenSource.Cancel();  
             this.logger.LogTrace("Disposing of staking loop.");
             this.stakingLoop?.Dispose();
-            this.stakingLoop = null;
             this.stakeCancellationTokenSource?.Dispose();
-            this.stakeCancellationTokenSource = null;
```

Comment:
why removing assignment to null? whoever created the code probably wanted it to be "cleaned" probably because it will actually cause that exception in case something is forgotton (otherwise it would just work improperly silently) - i'd say keep the logic, just fix the bug

Reply:
I get the feeling my dense brain is missing something obvious 😁 I'll do some reading up on CTS as haven't really worked much with it before.

Reply:
i'm aiming here for just minimal changes to the code, don't overthink it :-)

Reply:
After digging more into this, I'm not sure why the task is still running.  Calling Dispose() on the staking loop should have waited on the task so it should be completed.
```
       /// <summary>
        /// Wait for the loop task to complete.
        /// </summary>
        public void Dispose()
        {
            if (!this.RunningTask.IsCanceled)
            {
                this.logger.LogInformation("Waiting for {0} to finish.", this.Name);
                this.RunningTask.Wait();
            }
        }
```

Reply:
TBH my suggestion is we get rid of the disposal of the cancellationtokensource altogether from this code and let the garbage collector take care of it. This guy suggests that is the best way unless you are using the waithandle.  https://stackoverflow.com/questions/29779842/is-code-that-disposes-cancellationtokensource-while-tasks-are-canceling-correct

Reply:
As a matter of good practice, when something is disposable, we dispose it regardless of its internal implementation. Otherwise we would need to argue why we are not disposing it which is quite hard and could lead to errors.

I have checked the code and found out the design here is wrong and there is a race condition of the staking status, which is being controlled by interlocked operations. But let's discard this finding for this PR.

To finish with this PR, let's add questionmark after the cancellation token source in this code in stop staking and restore the line that assigned null to it.

If you want to dig deeper, let's make a Skype call.In that case please contact me on slack.